source/slang/slang-check-decl.cpp

```diff
@@ -8475,19 +8829,327 @@ namespace Slang
+            auto structVarDecl = m_astBuilder->create<VarDecl>();
+            addModifier(structVarDecl, m_astBuilder->create<LocalTempVarModifier>());
+            structVarDecl->type = TypeExp(structDeclType);
+            structVarDecl->initExpr = defaultConstructExpr;
+            structVarDecl->parentDecl = defaultInitListFunc;
+            structVarDecl->loc = seqStmt->loc;
+            ensureDecl(structVarDecl, DeclCheckState::VarInitExprAreChecked);
+
+            auto structVarDeclExpr = m_astBuilder->create<VarExpr>();
+            structVarDeclExpr->declRef = structVarDecl->getDefaultDeclRef();
+            structVarDeclExpr->name = structVarDecl->getName();
+            structVarDeclExpr->loc = structVarDecl->loc;
+            structVarDeclExpr->type = GetTypeForDeclRef(structVarDecl, structVarDecl->loc);
+            structVarDeclExpr->type.isLeftValue = true;
+            structVarDeclExpr->scope = block->scopeDecl->ownedScope;
+
+            auto structAssignExpr = m_astBuilder->create<AssignExpr>();
+            structAssignExpr->left = structVarDeclExpr;
+            structAssignExpr->right = defaultConstructExpr;
+            auto structAssignExprStmt = m_astBuilder->create<ExpressionStmt>();
+            structAssignExprStmt->expression = structAssignExpr;
+            structAssignExprStmt->loc = seqStmt->loc;
+            seqStmt->stmts.add(structAssignExprStmt);
+
+            // Part 2
+            // Insert `super::$DefaultInit()` into `this->$DefaultInit`
+            // Only insert if we have values to insert for
+            if (baseStructInfo)
+            {
+                if (auto baseStructDecl = as<StructDecl>(baseStructInfo->m_inheritanceBaseDecl))
+                {
+                    if (_structHasMemberWithValue(getASTBuilder(), structDecl))
+                    {
+                        Expr* defaultInitFunc = constructDefaultInitListFunc(this, baseStructDecl, baseStructInfo->m_inheritanceDecl->base.type, ConstructDefaultInitListOptions::PreferDefaultInitFunc);
+                        auto assign = m_astBuilder->create<AssignExpr>();
+                        assign->left = coerce(CoercionSite::Initializer, defaultInitFunc->type, structVarDeclExpr);
+                        assign->right = defaultInitFunc;
+                        auto stmt = m_astBuilder->create<ExpressionStmt>();
+                        stmt->expression = assign;
+                        stmt->loc = seqStmt->loc;
+                        seqStmt->stmts.add(stmt);
+                    }
+                }
+            }
+
+            // Part 3
+            // Assign initExpr to all members of structDecl
+            for (auto memberRef : membersOfStructDeclInstance)
+            {
+
+                auto member = memberRef.getDecl();
+                addCudaHostModifierIfRequired(defaultInitListFunc, member, foundCudaHostModifier);
+
+                auto initExpr = member->initExpr;
+                if (!initExpr)
```

Comment:
if !initExpr, and zero-init flag is set, we want to create a DefaultConstruct expr to create a 0-initialized value of member->type, and assign it to member.